### PR TITLE
[SPARK-12429][Streaming][Doc]Add Accumulator and Broadcast example for Streaming

### PR DESCRIPTION
**Pull request by zsxwing**
This PR adds Scala, Java and Python examples to show how to use Accumulator and Broadcast in Spark Streaming to support checkpointing.

### CODE REVIEW

**Comment by zsxwing** [comment]
@tdas could you take a look before I start to add Java and Python examples?

File: docs/streaming-programming-guide.md [hunk 1]
```diff
@@ -1415,6 +1415,95 @@ Note that the connections in the pool should be lazily created on demand and tim
 
 ***
 
+## Accumulator and Broadcast
+
+Accumulator and Broadcast cannot be recovered from checkpoint in Streaming. If you enable checkpoint and use Accumulator or Broadcast as well, you have to create lazily instantiated singleton instances for Accumulator and Broadcast so that they can be restarted on driver failures. This is shown in the following example.
```
Comment (BenFradet) [inline docs/streaming-programming-guide.md]: I'd say: "in **Spark** Streaming. If you enable checkpoint**ing** and use an Accumulator or Broadcast as well, you**'ll** have to create ..."

**Comment by zsxwing** [comment]
Added Java and Python examples.

**Comment by BenFradet** [comment]
lgtm

File: docs/streaming-programming-guide.md [hunk 2]
```diff
@@ -1415,6 +1415,185 @@ Note that the connections in the pool should be lazily created on demand and tim
 
 ***
 
+## Accumulator and Broadcast
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: Accumulators and Broadcast variables

File: docs/streaming-programming-guide.md [hunk 3]
```diff
@@ -1415,6 +1415,185 @@ Note that the connections in the pool should be lazily created on demand and tim
...
+
+Accumulator and Broadcast cannot be recovered from checkpoint in Spark Streaming. If you enable checkpointing and use Accumulator or Broadcast as well, you'll have to create lazily instantiated singleton instances for Accumulator and Broadcast so that they can be restarted on driver failures. This is shown in the following example.
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: I believe its called "Broadcast variable"

http://spark.apache.org/docs/latest/programming-guide.html#broadcast-variables

Might be a good idea to put links to the Spark programming guide

File: docs/streaming-programming-guide.md [hunk 4]
```diff
@@ -1415,6 +1415,185 @@ Note that the connections in the pool should be lazily created on demand and tim
...
+  val counts = rdd.filter { case (word, count) =>
+    if (blacklist.value.contains(word)) {
+      droppedWordsCounter += 1
+      false
+    } else {
+      true
+    }
+  }.collect().mkString("[", ", ", "]")
+  val output = "Counts at time " + time + " " + counts
+  println(output)
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: No need to give so much code. remove all the unnecessary cosmetic stuff like "mkString" and "print"  and Files.append ....

Same for all the other code.

File: docs/streaming-programming-guide.md [hunk 5]
```diff
@@ -1415,6 +1415,185 @@ Note that the connections in the pool should be lazily created on demand and tim
...
+        }
+      }
+    }
+    instance
+  }
+}
+
+object DroppedWordsCounter {
+
+  @volatile private var instance: Accumulator[Long] = null
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: For scala, cant this whole thing be replaced with lazy val?

File: docs/streaming-programming-guide.md [hunk 6]
```diff
@@ -1415,6 +1415,185 @@ Note that the connections in the pool should be lazily created on demand and tim
...
+          droppedWordsCounter.add(wordCount._2());
+          return false;
+        } else {
+          return true;
+        }
+      }
+    }).collect().toString();
+    String output = "Counts at time " + time + " " + counts;
+    System.out.println(output);
+    System.out.println("Dropped " + droppedWordsCounter.value() + " word(s) totally");
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: remove unnecessary stuff.

File: docs/streaming-programming-guide.md (hunk 5 above)
Comment (zsxwing) [inline docs/streaming-programming-guide.md]: Yes. Good point.

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaRecoverableNetworkWordCount.java [hunk 7]
```diff
@@ -41,7 +46,48 @@
 import org.apache.spark.streaming.api.java.JavaStreamingContextFactory;
 
 /**
- * Counts words in text encoded with UTF8 received from the network every second.
+ * Use this singleton to get or register `Broadcast`.
```
Comment (tdas) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaRecoverableNetworkWordCount.java]: "broadcast variable".
why is Broadcast in single quotes?

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaRecoverableNetworkWordCount.java [hunk 8]
```diff
@@ -41,7 +46,48 @@
...
+          instance = jsc.broadcast(wordBlacklist);
+        }
+      }
+    }
+    return instance;
+  }
+}
+
+/**
+ * Use this singleton to get or register `Accumulator`.
```
Comment (tdas) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaRecoverableNetworkWordCount.java]: Similar comment as above for Broadcast.. "an accumulator"

File: examples/src/main/scala/org/apache/spark/examples/streaming/RecoverableNetworkWordCount.scala [hunk 9]
```diff
@@ -23,13 +23,55 @@ import java.nio.charset.Charset
...
-import org.apache.spark.SparkConf
+import org.apache.spark.{Accumulator, SparkConf, SparkContext}
+import org.apache.spark.broadcast.Broadcast
 import org.apache.spark.rdd.RDD
 import org.apache.spark.streaming.{Time, Seconds, StreamingContext}
 import org.apache.spark.util.IntParam
 
 /**
- * Counts words in text encoded with UTF8 received from the network every second.
+ * Use this singleton to get or register `Broadcast`.
```
Comment (tdas) [inline examples/src/main/scala/org/apache/spark/examples/streaming/RecoverableNetworkWordCount.scala]: Same comments as above.

File: examples/src/main/scala/org/apache/spark/examples/streaming/RecoverableNetworkWordCount.scala [hunk 10]
```diff
@@ -23,13 +23,55 @@ import java.nio.charset.Charset
...
+    instance
+  }
+}
+
+/**
+ * Use this singleton to get or register `Accumulator`.
+ */
+object DroppedWordsCounter {
+
+  @volatile private var instance: Accumulator[Long] = null
```
Comment (tdas) [inline examples/src/main/scala/org/apache/spark/examples/streaming/RecoverableNetworkWordCount.scala]: same comment as above.

**Comment by tdas** [comment]
Small comments, other LGTM.

File: docs/streaming-programming-guide.md (hunk 5 above)
Comment (zsxwing) [inline docs/streaming-programming-guide.md]: Oh, no. We need `sc` here. So it cannot be a val.

File: docs/programming-guide.md [hunk 11]
```diff
@@ -806,7 +806,7 @@ However, in `cluster` mode, what happens is more complicated, and the above may
 
 What is happening here is that the variables within the closure sent to each executor are now copies and thus, when **counter** is referenced within the `foreach` function, it's no longer the **counter** on the driver node. There is still a **counter** in the memory of the driver node but this is no longer visible to the executors! The executors only see the copy from the serialized closure. Thus, the final value of **counter** will still be zero since all operations on **counter** were referencing the value within the serialized closure.  
 
-To ensure well-defined behavior in these sorts of scenarios one should use an [`Accumulator`](#AccumLink). Accumulators in Spark are used specifically to provide a mechanism for safely updating a variable when execution is split up across worker nodes in a cluster. The Accumulators section of this guide discusses these in more detail.
```
Comment (zsxwing) [inline docs/programming-guide.md]: I also fixed the broken `Accumulator` link in `programming-guide.md`

Comment (tdas) [inline docs/programming-guide.md]: nice!

**Pull request closed without merging**